optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/ListChangeMove.java

```diff
@@ -0,0 +1,103 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.heuristic.selector.move.generic.list;
+
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.api.score.director.ScoreDirector;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.heuristic.move.AbstractMove;
+import org.optaplanner.core.impl.score.director.InnerScoreDirector;
+
+/**
+ * Moves an element of a list planning variable. The moved element is identified by an entity instance and a position
+ * in that entity's list variable. The element is inserted at the given index in the given destination entity's list
+ * variable.
+ * <p>
+ * The move can be undone by simply flipping the source and destination entity+index.
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ */
+public class ListChangeMove<Solution_> extends AbstractMove<Solution_> {
+
+    private final Object sourceEntity;
+    private final int sourceIndex;
+    private final Object destinationEntity;
+    private final int destinationIndex;
+
+    private final ListVariableDescriptor<Solution_> variableDescriptor;
+
+    public ListChangeMove(
+            Object sourceEntity, int sourceIndex,
+            Object destinationEntity, int destinationIndex,
+            ListVariableDescriptor<Solution_> variableDescriptor) {
+        this.sourceEntity = sourceEntity;
+        this.sourceIndex = sourceIndex;
+        this.destinationEntity = destinationEntity;
+        this.destinationIndex = destinationIndex;
+        this.variableDescriptor = variableDescriptor;
+    }
+
+    @Override
+    protected AbstractMove<Solution_> createUndoMove(ScoreDirector<Solution_> scoreDirector) {
+        return new ListChangeMove<>(destinationEntity, destinationIndex, sourceEntity, sourceIndex, variableDescriptor);
+    }
+
+    @Override
+    protected void doMoveOnGenuineVariables(ScoreDirector<Solution_> scoreDirector) {
+        InnerScoreDirector<Solution_, ?> innerScoreDirector = (InnerScoreDirector<Solution_, ?>) scoreDirector;
+        // Remove element from the source entity.
+        innerScoreDirector.beforeVariableChanged(variableDescriptor, sourceEntity);
+        Object element = variableDescriptor.removeElement(sourceEntity, sourceIndex);
+        innerScoreDirector.afterVariableChanged(variableDescriptor, sourceEntity);
+        // Add element to the destination entity.
+        innerScoreDirector.beforeVariableChanged(variableDescriptor, destinationEntity);
+        variableDescriptor.addElement(destinationEntity, destinationIndex, element);
```

Comment:
presume destinationEntity == sourceEntity

input: Ann[A0, A1, A2, A3, A4, A5]
change move(Ann, 2, Ann, 4)
result A): Ann[A0, A1, A3, A4, A2, A5]
result B): Ann[A0, A1, A3, A2, A4, A5] because index 4 is before A4

Reply:
```
Given [A0, A1, A2, A3, A4, A5]
change move(Ann, 2, Ann, 0) => A) [A2, A0, A1, A3, A4, A5]  - B) [A2, A0, A1, A3, A4, A5]
change move(Ann, 2, Ann, 1) => A) [A0, A2, A1, A3, A4, A5]  - B) [A0, A2, A1, A3, A4, A5]
change move(Ann, 2, Ann, 2) => A) undoable                          - B) undoable
change move(Ann, 2, Ann, 3) => A) [A0, A1, A3, A2, A4, A5]  - B) undoable
change move(Ann, 2, Ann, 4) => A) [A0, A1, A3, A4, A2, A5]  - B) [A0, A1, A3, A2, A4, A5]
change move(Ann, 2, Ann, 5) => A) [A0, A1, A3, A4, A5, A2]  - B) [A0, A1, A3, A4, A2, A5]
change move(Ann, 2, Ann, 6) => A) undoable                          - B) [A0, A1, A3, A4, A5, A2]
```

Reply:
ChangeMove constructor needs a javadoc to specify the behavior if destinationEntity == sourceEntity